src/client/common/alias_dict.cpp

```diff
@@ -0,0 +1,188 @@
+        modified = true;
+        return true;
+    }
+
+    return false;
+}
+
+void mp::AliasDict::remove_aliases_for_instance(const std::string& instance)
+{
+    for (auto it = aliases.begin(); it != aliases.end();)
+    {
+        if (it->second.instance == instance)
+        {
+            modified = true;
+            it = aliases.erase(it);
+        }
+        else
+            ++it;
+    }
+}
+
+mp::optional<mp::AliasDefinition> mp::AliasDict::get_alias(const std::string& alias) const
+{
+    try
+    {
+        return aliases.at(alias);
+    }
+    catch (const std::out_of_range&)
+    {
+        return mp::nullopt;
+    }
+}
+
+void mp::AliasDict::load_dict()
+{
+    QFile db_file{QString::fromStdString(aliases_file)};
+
+    aliases.clear();
+
+    if (db_file.exists())
+    {
+        if (!db_file.open(QIODevice::ReadOnly))
+            throw std::runtime_error(fmt::format("Error opening file '{}'", db_file.fileName()));
+    }
+    else
+        return;
+
+    QJsonParseError parse_error;
+    QJsonDocument doc = QJsonDocument::fromJson(db_file.readAll(), &parse_error);
+    if (doc.isNull())
+        return;
+
+    QJsonObject records = doc.object();
+    if (records.isEmpty())
+        return;
+
+    for (auto it = records.constBegin(); it != records.constEnd(); ++it)
+    {
+        auto alias = it.key().toStdString();
+        auto record = it.value().toObject();
+        if (record.isEmpty())
+            break;
+
+        auto instance = record["instance"].toString().toStdString();
+        auto command = record["command"].toString().toStdString();
+
+        aliases.emplace(alias, mp::AliasDefinition{instance, command});
+    }
+
+    db_file.close();
+}
+
+void mp::AliasDict::save_dict()
+{
+    auto alias_to_json = [](const mp::AliasDefinition& alias) -> QJsonObject
+    {
+        QJsonObject json;
+        json.insert("instance", QString::fromStdString(alias.instance));
+        json.insert("command", QString::fromStdString(alias.command));
+
+        return json;
+    };
+
+    QJsonObject aliases_json;
+    for (const auto& record : aliases)
+    {
+        auto name = QString::fromStdString(record.first);
+        auto definition = alias_to_json(record.second);
+
+        aliases_json.insert(name, definition);
+    }
+
+    auto config_file_name = QString::fromStdString(aliases_file);
+
+    QTemporaryFile temp_file{mpu::create_temp_file_with_path(config_file_name)};
+
+    if (temp_file.open())
+    {
+        temp_file.setAutoRemove(false);
+
+        mp::write_json(aliases_json, temp_file.fileName());
+
+        temp_file.close();
+
+        if (MP_FILEOPS.exists(QFile{config_file_name}))
+        {
+            auto backup_file_name = config_file_name + ".bak";
+            QFile backup_file(backup_file_name);
+
+            if (MP_FILEOPS.exists(backup_file) && !MP_FILEOPS.remove(backup_file))
+                throw std::runtime_error(fmt::format("cannot remove old aliases backup file {}", backup_file_name));
```

Comment:
Should be able to cover this in a unit test, right?

Reply:
Done.